File: DAY 24/mergeSort.cpp
#include <iostream>
#include <vector>
using namespace std;

void merge(vector<int>& arr, int low, int mid, int high) {
    int left = low;
    int right = mid + 1;
    vector<int> temp;

    // Merge the two halves
    while (left <= mid && right <= high) {
        if (arr[left] <= arr[right]) {
            temp.push_back(arr[left++]);
        } else {
            temp.push_back(arr[right++]);
        }
    }

    // Remaining elements of left half
    while (left <= mid) {
        temp.push_back(arr[left++]);
    }

    // Remaining elements of right half
    while (right <= high) {
        temp.push_back(arr[right++]);
    }

    // Copy sorted temp back into original array
    for (int i = 0; i < temp.size(); i++) {
        arr[low + i] = temp[i];
    }
}

void mergeSort(vector<int>& arr, int low, int high) {
    if (low >= high) return;

    int mid = (low + high) / 2;

    mergeSort(arr, low, mid);
    mergeSort(arr, mid + 1, high);
    merge(arr, low, mid, high);
}

int main() {
    vector<int> arr = {4, 2, 1, 6};

    mergeSort(arr, 0, arr.size() - 1);

    cout << "Sorted array: ";
    for (int num : arr) {
        cout << num << " ";
    }
    cout << endl;

    return 0;
}

File: DAY 24/mergeSort.rs
fn mergeSort (arr: &mut [i32]){
    let len = arr.len();
    if len <=1 {
        return;
    }
    let mid = len/2;



    
    let mut left = arr[0..mid].to_vec();
    let mut right = arr[mid..len].to_vec();

        mergeSort(&mut arr[0..mid]);
    mergeSort(&mut arr[mid..len]);
    merge (&mut left, &mut right, arr);
}

fn merge(left: &mut [i32], right: &mut [i32], arr: &mut [i32]){
    let mut i = 0;
    let mut j = 0;
    let mut k =0;

    while i < left.len()  && j < right.len(){
        if left[i] <= right[i]{
            arr[k] = left[i];
            i+=1;
        } else{
            arr[k] = right[j];
            j+=1;
        }
        k+=1;
    }
    while i< left.len(){
        arr[k] = left[i];
        i+=1;
        k+-1
    }
    while j < right.len(){
        arr[k] = right[j];
        j+=1;
        k+=1;
    }
}

